brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -6,81 +6,145 @@ import com.twitter.algebird._
 
 import AnnotatedTree.AnnotatedTreeTraversal
 
+
+//map with a reservoir of up to `capacity` randomly chosen keys
+case class SampledMap[A,B](capacity: Int) {
+  var mapValues = Map[A,B]()
+  var randValues = Map[A,Double]()
+  var threshold = 0.0
+  val rand = new util.Random
+
+  private def randValue(key: A): Double = {
+    randValues.get(key) match {
+      case Some(r) => r
+      case None => {
+        val r = rand.nextDouble
+        randValues += key->r
+
+        if(randValues.size <= capacity && r >= threshold)
+          threshold = r
+        else if(randValues.size > capacity && r < threshold) {
+          println("evicting")
+          val bottomK = randValues.toList.sortBy{_._2}.take(capacity)
+          val keep = bottomK.map{_._1}.toSet
+          threshold = bottomK.last._2
+          mapValues = mapValues.filterKeys(keep)
+        }
+
+        r
+      }
+    }
+  }
+
+  def containsKey(key: A): Boolean = randValue(key) <= threshold
+  def update(key: A, value: B) {
+    if(containsKey(key))
+      mapValues += key -> value
```

Comment:
Since we're just mutating this structure in place why don't we use `mutable.Map` for better performance?